### DISCUSSION TITLE
How to manually set the floorplan for each processing element (PE)

### DISCUSSION THREAD

**Discussion by yyan7223** [The-OpenROAD-Project/OpenROAD-flow-scripts#2482] [Q&A] [answered]
Hi alls,

I use OpenRoad to automatically generate the layout of a 2x2 PE array, 4 PEs are not evenly distributed in the die:
![layout2x2](https://github.com/user-attachments/assets/3395e4b2-5f3a-4b75-bf0f-53233fae8cbd)

Is there any tutorial about what I should do if I want to make it look like this:
<img width="227" alt="ideal" src="https://github.com/user-attachments/assets/decffee7-5d04-40fb-b335-90c1ea7a12c7">

Best Regards,
Yufei

**Comment by maliberty** [comment]
Are the PEs macros or gates?  Why is symmetry requried?  You should consider constraining the IO locations as well.

**Reply by yyan7223** [reply]
Thanks for the reply, I'm new to the backend.

PE array is an accelerator prototype I designed to accerate CNN. Each PE contains crossbar, Functional Unit (Arithmetic and logic operations), and many registers. So they should be gates I think because they are small and simple. (Macro should be some large modules like DRAM, SRAM, or some complex processor cores?)

Well symmetry is not necessary, I simply just think if I can specify the floorplan by myself. I'm doing a thermal simulation which requires the floorplan and dynamic power traces files like shown in the attached file.
[floorplan.txt](https://github.com/user-attachments/files/17450375/floorplan.txt)
[powertrace.txt](https://github.com/user-attachments/files/17450376/powertrace.txt)

So is it possible for OpenRoad to generate the floorplan of each module within the PE, as well as their power traces when running a specific benchmark? Manually parsing from `flow/reports` is also fine, could you please provide me with a top guidance? Appreciate that.

Best regards,
Yufei

**Comment by maliberty** [comment, marked as answer]
You could create placement regions an assign gates to each region.  Internally the regions need not be symmetric but you could control their overall location.  If you want perfect symmetry then you'll need to make a block.

For power you would need to generate a VCD file to represent a specific benchmark.
